pkg/syncer/metadatasyncer.go

```diff
@@ -3391,5 +3391,91 @@ func cnsVolumeInfoCRUpdated(oldObj interface{}, newObj interface{}, metadataSync
+		return
+	}
+	cnsOperatorClient, err := k8s.NewClientForGroup(ctx, restConfig, cnsoperatorv1alpha1.GroupName)
+	if err != nil {
+		log.Errorf("cnsVolumeInfoCRUpdated: Failed to create CnsOperator client. Err: %+v", err)
+		return
+	}
+
+	storagePolicyUsageInstanceName := newCnsVolumeInfoObj.Spec.StorageClassName + "-" +
+		storagepolicyv1alpha1.NameSuffixForSnapshot
+	storagePolicyUsageCR := &storagepolicyv1alpha1.StoragePolicyUsage{}
+	err = cnsOperatorClient.Get(ctx, k8stypes.NamespacedName{
+		Namespace: newCnsVolumeInfoObj.Spec.Namespace,
+		Name:      storagePolicyUsageInstanceName},
+		storagePolicyUsageCR)
+	if err != nil {
+		log.Errorf("failed to fetch %s instance with name %q from supervisor namespace %q. Error: %+v",
+			storagepolicyv1alpha1.CRDSingular, storagePolicyUsageInstanceName,
+			newCnsVolumeInfoObj.Spec.Namespace, err)
+		return
+	}
+
+	patchedStoragePolicyUsageCR := storagePolicyUsageCR.DeepCopy()
+	if !newCnsVolumeInfoObj.Spec.ValidAggregatedSnapshotSize {
+		// If ValidAggregatedSnapshotSize field of new CnsVolumeInfo CR is false, then add annotation
+		// "csi.vsphere.missing-snapshot-aggregated-capacity: true" on StoragePolicyUsage CR.
+		log.Infof("Couldn't retrieve aggregated snapshot size for volume %q, adding annotation %s on "+
+			"StoragePolicyUsage CR", common.MissingSnapshotAggregatedCapacity, newCnsVolumeInfoObj.Spec.VolumeID)
+		patchAnnotation := common.MergeMaps(newCnsVolumeInfoObj.Annotations,
+			map[string]string{common.MissingSnapshotAggregatedCapacity: "true"})
+		patchedStoragePolicyUsageCR.Annotations = patchAnnotation
+	} else {
+		if newCnsVolumeInfoObj.Spec.AggregatedSnapshotSize != nil &&
+			storagePolicyUsageCR.Status.ResourceTypeLevelQuotaUsage != nil &&
+			storagePolicyUsageCR.Status.ResourceTypeLevelQuotaUsage.Used != nil {
+			// Delete annotation "csi.vsphere.missing-snapshot-aggregated-capacity: true" on StoragePolicyUsage CR
+			// if it was present, since ValidAggregatedSnapshotSize is true.
+			delete(patchedStoragePolicyUsageCR.Annotations, common.MissingSnapshotAggregatedCapacity)
+
+			var oldAggregatedSnapshotSize int64
+			if oldCnsVolumeInfoObj.Spec.AggregatedSnapshotSize != nil {
+				oldAggregatedSnapshotSize = oldCnsVolumeInfoObj.Spec.AggregatedSnapshotSize.Value()
+			} else {
+				oldAggregatedSnapshotSize = 0
+			}
+			// Update the StoragePolicyUsage "Used" field based on old and new AggregatedSnapshotSize
+			if newCnsVolumeInfoObj.Spec.AggregatedSnapshotSize.Value() > oldAggregatedSnapshotSize {
+				// Aggregated snapshot size increased
+				aggregatedSnapSizeDifference := newCnsVolumeInfoObj.Spec.AggregatedSnapshotSize.Value() -
+					oldAggregatedSnapshotSize
+				snapSizeDifference := resource.NewQuantity(aggregatedSnapSizeDifference, resource.BinarySI)
+				patchedStoragePolicyUsageCR.Status.ResourceTypeLevelQuotaUsage.Used.Add(*snapSizeDifference)
+				log.Infof("cnsVolumeInfoCRUpdated: aggregated snapshot size increased by %v bytes, increase used "+
```

Comment:
This log indicates that it's already done, which could be misleading if the patch fails.

Reply:
Updated it accordingly